src/test/java/bio/terra/workspace/integration/WorkspaceIntegrationTest.java

```diff
@@ -28,24 +29,94 @@
+  //  different class files based on the the type of workspace action (Create, Get, Delete, etc).
 
+  @Autowired private WorkspaceManagerTestClient workspaceManagerTestClient;
   @Autowired private TestUtils testUtils;
-
   @Autowired private TestConfiguration testConfig;
-
   @Autowired private AuthService authService;
+  private static final Logger logger = LoggerFactory.getLogger(WorkspaceIntegrationTest.class);
+  private final ConcurrentHashMap<String, List<UUID>> testToWorkspaceIdsMap =
+      new ConcurrentHashMap<>();
+
+  // TODO: Create custom annotations for this tag, and also for the "integration" and "unit" tags in
+  //  this test and elsewhere. AS-431.
+  private final String TAG_NEEDS_CLEANUP = "needs-cleanup";
 
   @BeforeEach
   public void setup() {}
 
+  /*
+   * TODO: Cloud Resource Library (CRL) is planning to broadly address resource cleanup for integration tests in MC Terra
+   *  applications.
+   *  The doc below proposes a "Resource Tracking" approach for cleaning up. We should follow the developments in CRL for
+   *  potential utilization here in Workspace Manager.
+   *  Doc: https://docs.google.com/document/d/13mYVJML_fOLsX1gxQxRJgECUNT27dAMKzEJxUEvtQqM/edit#heading=h.x06ofvfgp7wt
+   */
+  @AfterEach
+  public void tearDown(TestInfo testInfo) throws Exception {
+    Set<String> tags = testInfo.getTags();
+    if (tags != null && tags.contains(TAG_NEEDS_CLEANUP)) {
+      String testName = testInfo.getDisplayName();
+      cleanUpWorkspaces(testToWorkspaceIdsMap.get(testName));
+    }
+  }
+
+  @Test
+  @Tag(TAG_NEEDS_CLEANUP)
+  public void createWorkspace(TestInfo testInfo) throws Exception {
+    UUID workspaceId = UUID.randomUUID();
+    testToWorkspaceIdsMap.put(testInfo.getDisplayName(), Collections.singletonList(workspaceId));
+
+    WorkspaceResponse<CreatedWorkspace> workspaceResponse = createDefaultWorkspace(workspaceId);
+
+    Assertions.assertEquals(workspaceResponse.getStatusCode(), HttpStatus.OK);
+    Assertions.assertTrue(workspaceResponse.isResponseObject());
+    CreatedWorkspace createdWorkspace = workspaceResponse.getResponseObject();
+    Assertions.assertEquals(workspaceId.toString(), createdWorkspace.getId());
+  }
+
+  @Test
+  public void deleteWorkspace() throws Exception {
+    UUID workspaceId = UUID.randomUUID();
+    WorkspaceResponse<CreatedWorkspace> workspaceResponse = createDefaultWorkspace(workspaceId);
+
+    String userEmail = testConfig.getServiceAccountEmail();
+    String token = authService.getAuthToken(userEmail);
+    String path = testConfig.getWsmWorkspacesBaseUrl() + "/" + workspaceId;
+    DeleteWorkspaceRequestBody body = new DeleteWorkspaceRequestBody().authToken(token);
+    String jsonBody = testUtils.mapToJson(body);
+
+    WorkspaceResponse<?> deleteWorkspaceResponse =
+        workspaceManagerTestClient.delete(userEmail, path, jsonBody);
+
+    Assertions.assertEquals(deleteWorkspaceResponse.getStatusCode(), HttpStatus.NO_CONTENT);
+  }
```

Comment:
should probably mark this one as needing cleanup too. If this test creates a workspace, but then fails to delete it, the workspace will be orphaned and need cleanup. Potentially, if the final assertion succeeds (i.e. the test passes), you could remove the entry in the cleanup map so the tearDown doesn't try to delete it a second time.

Reply:
My initial thought was: The 'cleanup' does the same thing as the deleteWorkspace step in this test (i.e. they both call the same http endpoint with the same user, path, etc). So if the deleteWorkspace step fails, the cleanup is likely to fail (unless it's a non-persistent issue such as temporary unavailability, timeout, etc).

I can remove the entry and proceed to clean up for this test. However, I'm wondering if the things a test will need to know/do are beginning to add up. Currently, a test (code/human) that needs cleanup will need to know to (1) use the `@Tag(TAG_NEEDS_CLEANUP)` annotation, and (2) add the the Id's to the map. Maybe there's a way we can make the preparation for cleanup a little more seamless. I'll think of ideas, but don't have any right now. Also, the development of Janitor service _should_ be able to help with this.

Let me know what you think.